src/ethereum/index.ts

```diff
@@ -0,0 +1,871 @@
+        value: Number(amountWei),
+        gasLimit: this.calculateGasLimit(gasConsumption),
+        data: this.contractPool.methods.stake(source).encodeABI(),
+      };
+    } catch (error) {
+      throw this.handleError('STAKE_ERROR', error);
+    }
+  }
+
+  /**
+   * Unstake value from active autocompound balance.
+   * AllowedInterchangeNum is max allowed number interchanges with pending stakers.
+   * Unstaked immediately if value <= pool pending balance or create withdraw request.
+   * Interchange disallowed as default.
+   *
+   * @param address - Sender address.
+   * @param amount - Unstake amount in ETH.
+   * @param allowedInterchangeNum - Max allowed number of interchanges. Default is 0.
+   * @param source - Unstake source. Default is '0'.
+   *
+   * @returns A Promise that resolves to the unsigned ETH transaction object.
+   *
+   * @throws Will throw an Error if the amount is not a string, the balance is less than the amount, or the contract call fails.
+   */
+  public async unstake(
+    address: string,
+    amount: string,
+    allowedInterchangeNum: number = 0,
+    source: string = '0',
+  ): Promise<Transaction> {
+    if (typeof amount !== 'string') {
+      this.throwError('WRONG_TYPE_MESSAGE');
+    }
+
+    try {
+      const balance = await this.autocompoundBalanceOf(address);
+      // Check for type overflow
+      if (allowedInterchangeNum > UINT16_MAX) {
+        allowedInterchangeNum = UINT16_MAX;
+      }
+
+      if (balance.lt(new BigNumber(amount))) {
+        this.throwError('MAX_AMOUNT_FOR_UNSTAKE_ERROR', balance.toString());
+      }
+
+      const amountWei = this.web3.utils.toWei(amount, 'ether');
+      const gasConsumption = await this.contractPool.methods
+        .unstake(amountWei, allowedInterchangeNum, source)
+        .estimateGas({ from: address });
+
+      // Create the transaction
+      return {
+        from: address,
+        value: 0,
+        to: this.addressContractPool,
+        gasLimit: this.calculateGasLimit(gasConsumption),
+        data: this.contractPool.methods
+          .unstake(amountWei, allowedInterchangeNum, source)
+          .encodeABI(),
+      };
+    } catch (error) {
+      throw this.handleError('UNSTAKE_ERROR', error);
+    }
+  }
+
+  /**
+   * Simulate unstake transaction and return amount of instant unstake.
+   * Required to compare evaluation of allowedInterchangeNum parameter.
+   *
+   * @param address - Sender address.
+   * @param amount - Unstake amount in ETH.
+   * @param allowedInterchangeNum - Max allowed number of interchanges. Default is 1.
+   * @param source - Unstake source. Default is '0'.
+   *
+   * @returns A Promise that resolves to a BigNumber representing the instant unstake amount in ETH.
+   *
+   * @throws Will throw an Error if the balance is less than the amount or the contract call fails.
+   */
+  public async simulateUnstake(
+    address: string,
+    amount: string,
+    allowedInterchangeNum: number = 1,
```

Comment:
Why do `unstake` and `simulateUnstake` have different defaults? if they should be the same, I'd suggest creating a common constant for that.